### ISSUE TITLE
Remove vender code from repository

### ISSUE THREAD

**Issue by alerque** [topanswers/topanswers#3]
One of the first things I noticed whet this code was posted was that a number of vendor ed dependencies were directly added to the base repository. This is a big no-no for a number of reasons. It seems innocuous at first and eases deployment (since any git checkout is ready to deploy) but this will be a nightmare down the road as the repository starts dragging along baggage.

All external dependencies should be vendored and a local setup script should be used to fetch and configure that stuff. For the dependencies so far `npm` is probably the right tool: a `package.json` file can list the dependencies (and _exact_ versions) required and `npm install` on a checkout will get everything running. (Another approach is to keep a branch or repo that does have _deployed_ versions of the code with vendor stuff backed in, but not use this as the master branch or even repository.)

I think the biggest decisions is actually whether or not to rebase the entire git history to strip vendored code. It isn't a lot of baggage at this point, but when you figure how many developers and years it will all be dragged around, it might actually be worth doing this rebase ASAP. If we decide to go that route I can help with the implementation. It will be an inconvenience to anybody that has already forked or checked out the code in that they will have to reset/rebase their branches to match in order to contribute. Some pain now and get it over with, or delay until the anoyances build up over time?

**Comment by jackdouglas** [comment]
> Some pain now and get it over with, or delay until the anoyances build up over time?

Short term pain and get it over with sounds better to me.

> I think the biggest decisions is actually whether or not to rebase the entire git history to strip vendored code.

Does this mean losing the commit history that we have already built up? If so that would be a shame, but not the end of the world.

> For the dependencies so far npm is probably the right tool

I didn't realise that npm could handle non-javascript dependencies like Fork Awesome. What about fonts we have downloaded from Font Library or other dependencies (now or in the future) that can't be satisfied via npm?

One obvious thing I'm keen not to break: we need to be able to make a library change to dev/test (like move to a different version of jQuery) and test those changes without affecting prod. It's OK to isolate those changes from other code changes — we don't need to be able to simultaneously change library and code as we can now. I just can't quite get my head round what tis will mean for our workflow.

**Comment by alerque** [comment]
> Does this mean losing the commit history that we have already built up?

No. It means using `git filter-branch` to drop the blobs from any commits where they happen to appear.

I'll work on this and post a parallel branch for review.

> What about ... other dependencies (now or in the future) that can't be satisfied via npm?

Goon question. My go-to tool for this is actually GNU-Make, which can be *made* to do just about anything. But for that matter so can `npm`, you can easily add a script in any language and call out to it part of the routine to fetch dependencies, build anything that needs built (for example CSS pre-processing, Javascript minification, Markdown pre-rendered to HTML, etc., and generally get the directory ready to be a deployment.

**Comment by alerque** [comment]
While I'm thinking of it, one minor "disadvantage" to keep in mind if something goes sideways @jackdouglas: when this happens, old commits in the history before the dependency install code get added will not be deployable any more. One would have to use the original unfiltered branch to revert if that were to be come necessary.

**Comment by jackdouglas** [comment]
> old commits in the history before the dependency install code get added will not be deployable any more

thanks — and I think it's unlikely that that will be an issue.
